Mathlib/Tactic/ProjectionNotation.lean

```diff
@@ -0,0 +1,112 @@
+register_option pp.collapseStructureProjections : Bool := {
+  defValue := true
+  group := "pp"
+  descr := "(pretty printer, Mathlib extension) display structure projections using field notation"
+}
+
+def getPPCollapseStructureProjections (o : Options) : Bool :=
+  o.get pp.structureProjections.name (!getPPAll o)
+
+/-- Like the projection delaborator from core Lean, but collapses projections to parent
+structures into a single projection.
+
+The only functional difference from `Lean.PrettyPrinter.Delaborator.delabProjectionApp` is
+the `walkUp` function. -/
+@[delab app]
+partial def delabProjectionApp' : Delab := whenPPOption getPPCollapseStructureProjections $ do
+  let e@(Expr.app fn _) ← getExpr | failure
+  let .const c@(.str _ f) _ := fn.getAppFn | failure
+  let env ← getEnv
+  let some info := env.getProjectionFnInfo? c | failure
+  -- can't use with classes since the instance parameter is implicit
+  guard <| !info.fromClass
+  -- projection function should be fully applied (#struct params + 1 instance parameter)
+  -- TODO: support over-application
+  guard <| e.getAppNumArgs == info.numParams + 1
+  -- If pp.explicit is true, and the structure has parameters, we should not
+  -- use field notation because we will not be able to see the parameters.
+  let expl ← getPPOption getPPExplicit
+  guard <| !expl || info.numParams == 0
+
+  /- Consume projections to parent structures. -/
+  let rec walkUp {α} (done : DelabM α) : DelabM α := withAppArg do
+    let (Expr.app fn _) ← getExpr | done
+    let .const c@(.str _ field) _ := fn.getAppFn | done
+    let some structName := env.getProjectionStructureName? c | failure
+    let some _ := isSubobjectField? env structName field | done
+    walkUp done
+
+  walkUp do
+    let appStx ← delab
+    `($(appStx).$(mkIdent f):ident)
+
+/--
+Defines an `app_unexpander` for the given function to support a basic form of projection
+notation. It is *only* for functions whose first explicit argument is the receiver
+of the generalized field notation. That is to say, it is only meant for transforming
+`C.f c x y z ...` to `c.f x y z ...` for `c : C`.
+
+It can be used to help get projection notation to work for function-valued structure fields,
+since the default projection delaborator cannot handle excess arguments.
+
+Example for generalized field notation:
+```
+structure A where
+  n : Nat
+
+def A.foo (a : A) (m : Nat) : Nat := a.n + m
+
+pp_extended_field_notation A.foo
+```
+Now, `A.foo x m` pretty prints as `x.foo m`. It also adds a rule that
+`A.foo x.toA m` pretty prints as `x.foo m`. This rule is meant to combine with
+the projection collapse delaborator, so that `A.foo x.toB.toA m` also will
+pretty print as `x.foo m`.
+
+Since this last rule is a purely syntactic transformation,
+it might lead to output that does not round trip, though this can only occur if
+there exists an `A`-valued `toA` function that is not a parent projection that
+happens to be pretty printable using dot notation.
```

Comment:
Would you mind adding an explicit example of what this scenario ("that happens to be pretty printable using dot notation") would look like.

I'm very excited to have this PR, but given there is a round-tripping gotcha I'd prefer it's extremely well documented. :-)

Reply:
Done!